### DISCUSSION TITLE
sunbird-rc-core - Developer Setup

### DISCUSSION THREAD

**Discussion by maheshkumargangula** [Sunbird-RC/community#106] [Issues] [answered]
@dileepbapat - I'm trying to setup sunbird-rc-core (registry) and run on my local machine. The configuration details are not clear. Can you please help me with the developer setup guide and a sample configuration to run the service.
cc: @vrayulu

**Comment by parthlawate** [comment]
Here is a guide done by @gamemaker1 https://github.com/gamemaker1/registry-cli/blob/main/docs/getting-started.md

**Reply by gamemaker1** [reply]
The installation and getting started guide was merged into the core repository yesterday - you can now find it [here](https://github.com/sunbird-rc/sunbird-rc-core/tree/main/docs).

**Comment by parthlawate** [comment]
I was wondering where that went :)  We are keeping all the documentation in the community Repo and that syncs to the Gitbook Documentation website.. Could you move it to the 'Setup the backend' section here https://github.com/Sunbird-RC/community/tree/main/use ? 

it will reflect here https://docs.sunbirdrc.dev/use/setup-the-backend 

Thanks for your contribution ! This guide is really helping :)

**Comment by parthlawate** [comment, marked as answer]
Also do feel free to improve the structure there..

**Reply by maheshkumargangula** [reply]
@parthlawate @gamemaker1 - The documentation explains how to use an existing docker image and setup the service or using registry-cli to setup. 

I want to compile and run the service from the source code. Please do the needful for the same.

**Reply by gamemaker1** [reply]
Yes you are right, that guide uses the `registry-cli` and docker images to setup a registry instance. I overlooked the word `developer` in the title 🤦🏾 

I will be creating an advanced installation guide (which also acts as a developer setup guide) soon that will go through setting up keycloak, and running the registry from the source (using java).

But in the meantime, let me try to give you a quick overview of the same:

1. Install Java 8, Docker, Docker Compose.
2. Clone the [`sunbird-rc-core`](https://github.com/sunbird-rc/sunbird-rc-core.git) repository.
3. Run the `configure-dependencies.sh` script in the root of the repo.
4. In the `/java` folder of the repository, run `./mvnw clean install -DskipTests`. This should create a JAR file in the `/java/registry/target` folder.
5. Place all your schema files (sample schemas for a simple teacher-student registry can be found [here](https://github.com/gamemaker1/registry-setup-files/tree/main/schemas/)) in the `/java/registry/src/main/resources/public/_schemas/` folder.
6. Download [this](https://github.com/gamemaker1/registry-setup-files/blob/main/docker-compose.yaml) Docker Compose file. Comment out the `volumes` part under the `kc` section in the file. Run `docker-compose up kc db es cs` to start Keycloak (`kc`), Postgres (`db`), Elastic Search (`es`) and the Claims Service (`cs`).
7. Edit your hosts file (`/etc/hosts`) and add the following line: `127.0.0.1 kc`
8. You need to configure keycloak (http://localhost:8080, user `admin` pass `admin`) (add a `sunbird-rc` realm, then create `admin-api` and `registry-frontend` clients. Allow service accounts for the `admin-api` client and give it the `manage-users` scope under `realm-management`).
9. Once keycloak has been configured, run the registry using `java -jar java/registry/target/registry.jar`.

Anytime you make a change to the registry's source, you need to recompile the jar (`./mvnw clean install -DskipTests`).

Feel free to post any questions here if you face any problems.

**Reply by parthlawate** [reply]
@gamemaker1 a gentle reminder to update this guide on the community repo :) getting a lot of demand for it !

**Reply by gamemaker1** [reply]
Hi @parthlawate,

I have taken a crack at creating documentation for developers. It includes:

- Installation and getting started guide (using the Registry CLI)
- Advanced (developer) getting started guide (build and run registry from source)
- API reference
- Contributing guide

I have temporarily hosted it [here](https://gamemaker1.github.io/sunbird-rc-dev-docs/) for everyone to preview. Could you let me know where I should add this in the community repository, I will open a PR with the documentation added in the specified folder under `sunbird-rc/community`.

Thanks,
Vedant

PS: Here is a screenshot of the documentation

![sunbird-rc-dev-docs](https://user-images.githubusercontent.com/34235681/138923942-87be202f-30ab-4719-adcc-29392dca9f60.png)

**Reply by parthlawate** [reply]
this looks great @gamemaker1 . You can submit the PR in the https://github.com/Sunbird-RC/community/tree/main/use folder.. Also feel free to modify the existing related files - remove/add /change that this guide will replace..
